src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.Value.cs

```diff
@@ -0,0 +1,244 @@
+                        }
+                    }
+                }
+
+                info = default;
+                return false;
+            }
+
+            private static string? SwitchVersionedPlatformName(string methodName)
+            => methodName switch
+            {
+                IsWindowsVersionAtLeast => Windows,
+                IsMacOSVersionAtLeast => MacOS,
+                IsIOSVersionAtLeast => IOS,
+                IsAndroidVersionAtLeast => Android,
+                IsFreeBSDVersionAtLeast => FreeBSD,
+                IsTvOSVersionAtLeast => TvOS,
+                IsWatchOSVersionAtLeast => WatchOS,
+                _ => null
+            };
+
+            private static string? SwitchPlatformName(string methodName)
+                 => methodName switch
+                 {
+                     IsWindows => Windows,
+                     IsLinux => Linux,
+                     IsMacOS => MacOS,
+                     IsIOS => IOS,
+                     IsBrowser => Browser,
+                     IsAndroid => Android,
+                     IsFreeBSD => FreeBSD,
+                     IsTvOS => TvOS,
+                     IsWatchOS => WatchOS,
+                     _ => null
+                 };
+
+            private static bool TryDecodeRuntimeInformationIsOSPlatform(
+                IOperation argumentValue,
+                INamedTypeSymbol osPlatformType,
+                [NotNullWhen(returnValue: true)] out string? osPlatformName)
+            {
+                if ((argumentValue is IPropertyReferenceOperation propertyReference) &&
+                    propertyReference.Property.ContainingType.Equals(osPlatformType))
+                {
+                    osPlatformName = propertyReference.Property.Name;
+                    return true;
+                }
+
+                osPlatformName = null;
+                return false;
+            }
+
+            private static bool TryDecodeOSVersion(
+                ImmutableArray<IArgumentOperation> arguments,
+                ValueContentAnalysisResult? valueContentAnalysisResult,
+                [NotNullWhen(returnValue: true)] out Version? osVersion,
+                int skip = 0)
+            {
+
+                using var versionBuilder = ArrayBuilder<int>.GetInstance(4, fillWithValue: 0);
+                var index = 0;
+
+                foreach (var argument in arguments.Skip(skip))
+                {
+                    if (!TryDecodeOSVersionPart(argument, valueContentAnalysisResult, out var osVersionPart))
+                    {
+                        osVersion = null;
+                        return false;
+                    }
+
+                    versionBuilder[index++] = osVersionPart;
+                }
+
+                osVersion = versionBuilder[3] == 0 ? versionBuilder[2] == 0 ? new Version(versionBuilder[0], versionBuilder[1]) :
+                     new Version(versionBuilder[0], versionBuilder[1], versionBuilder[2]) :
```

Comment:
Not: This feels hard to read. Consider moving to a local function with use of explicit if statements.